ee/packages/federation-matrix/src/api/_matrix/media.ts

```diff
@@ -0,0 +1,185 @@
+const isBufferResponseProps = ajv.compile(BufferResponseSchema);
+
+const SECURITY_HEADERS = {
+	'X-Content-Type-Options': 'nosniff',
+	'X-Frame-Options': 'DENY',
+	'Content-Security-Policy': "default-src 'none'; img-src 'self'; media-src 'self'",
+	'Strict-Transport-Security': 'max-age=31536000; includeSubDomains',
+};
+
+function createMultipartResponse(
+	buffer: Buffer,
+	mimeType: string,
+	fileName: string,
+	metadata: Record<string, any> = {},
+): { body: Buffer; contentType: string } {
+	const boundary = crypto.randomBytes(16).toString('hex');
+	const parts: string[] = [];
+
+	parts.push(`--${boundary}`, 'Content-Type: application/json', '', JSON.stringify(metadata));
+	parts.push(`--${boundary}`, `Content-Type: ${mimeType}`, `Content-Disposition: attachment; filename="${fileName}"`, '');
+
+	const headerBuffer = Buffer.from(`${parts.join('\r\n')}\r\n`);
+	const endBoundary = Buffer.from(`\r\n--${boundary}--\r\n`);
+
+	return {
+		body: Buffer.concat([headerBuffer, buffer, endBoundary]),
+		contentType: `multipart/mixed; boundary=${boundary}`,
+	};
+}
+
+async function getMediaFile(mediaId: string, serverName: string): Promise<{ file: IUpload; buffer: Buffer } | null> {
+	const file = await MatrixMediaService.getLocalFileForMatrixNode(mediaId, serverName);
+	if (!file) {
+		return null;
+	}
+
+	const buffer = await MatrixMediaService.getLocalFileBuffer(file);
+	return { file, buffer };
+}
+
+export const getMatrixMediaRoutes = (homeserverServices: HomeserverServices) => {
+	const { config } = homeserverServices;
+	const router = new Router('/federation');
+
+	router.get(
+		'/v1/media/download/:mediaId',
+		{
+			params: isMediaDownloadParamsProps,
+			response: {
+				200: isBufferResponseProps,
+				401: isErrorResponseProps,
+				404: isErrorResponseProps,
+				429: isErrorResponseProps,
+				500: isErrorResponseProps,
+			},
+			tags: ['Federation', 'Media'],
+		},
+		async (c) => {
+			try {
+				const { mediaId } = c.req.param();
+				const { serverName } = config;
+
+				const result = await getMediaFile(mediaId, serverName);
+				if (!result) {
+					return {
+						statusCode: 404,
+						body: { errcode: 'M_NOT_FOUND', error: 'Media not found' },
+					};
+				}
+
+				const { file, buffer } = result;
+
+				const mimeType = file.type || 'application/octet-stream';
+				const fileName = file.name || mediaId;
+
+				const multipartResponse = createMultipartResponse(buffer, mimeType, fileName);
+
+				return {
+					statusCode: 200,
+					headers: {
+						...SECURITY_HEADERS,
+						'content-type': multipartResponse.contentType,
+						'content-length': String(multipartResponse.body.length),
+					},
+					body: multipartResponse.body,
+				};
+			} catch (error) {
+				logger.error('Federation media download error:', error);
+				return {
+					statusCode: 500,
+					body: { errcode: 'M_UNKNOWN', error: 'Internal server error' },
+				};
+			}
+		},
+	);
+
+	router.get(
+		'/v1/media/thumbnail/:mediaId',
```

Comment:
Since we do not generate thumbs I'd reply as 400 as per spec:
```
{
  "errcode": "M_UNKNOWN",
  "error": "Cannot generate thumbnails for the requested content"
}
```
Better to not return a content than return the full file as a thumb